db/commitlog/commitlog.cc

```diff
@@ -1392,83 +1393,84 @@ void db::commitlog::segment_manager::flush_segments(uint64_t size_to_remove) {
-                    static constexpr size_t buf_size = 4 * segment::alignment;
-                    size_t zerofill_size = max_size - align_down(existing_size, segment::alignment);
-                    return do_with(allocate_single_buffer(buf_size), zerofill_size, [this, f](temporary_buffer<char>& buf, uint64_t& rem) mutable {
-                        std::fill(buf.get_write(), buf.get_write() + buf.size(), 0);
-                        return repeat([this, f, &rem, &buf]() mutable {
-                            if (rem == 0) {
-                                return make_ready_future<stop_iteration>(stop_iteration::yes);
-                            }
-                            static constexpr size_t max_write = 128 * 1024;
-                            auto n = std::min(max_write / buf_size, 1 + rem / buf_size);
-
-                            std::vector<iovec> v;
-                            v.reserve(n);
-                            size_t m = 0;
-                            while (m < rem && n < max_write) {
-                                auto s = std::min(rem - m, buf_size);
-                                v.emplace_back(iovec{ buf.get_write(), s});
-                                m += s;
-                            }
-                            return f.dma_write(max_size - rem, std::move(v), service::get_local_commitlog_priority()).then([&rem](size_t s) mutable {
-                                rem -= s;
-                                return stop_iteration::no;
-                            });
-                        });
-                    });
-                });
-            });
+
+                // re-open without o_dsync for pre-alloc
+                co_await f.close();
+                f = co_await open_file_dma(filename, flags & open_flags(~int(open_flags::dsync)), opt);
+                co_await f.allocate(existing_size, max_size - existing_size);
+
+                static constexpr size_t buf_size = 4 * segment::alignment;
+                size_t zerofill_size = max_size - align_down(existing_size, segment::alignment);
+                auto rem = zerofill_size;
+
+                auto buf = allocate_single_buffer(buf_size);
+                while (rem != 0) {
+                    static constexpr size_t max_write = 128 * 1024;
+                    auto n = std::min(max_write / buf_size, 1 + rem / buf_size);
+
+                    std::vector<iovec> v;
+                    v.reserve(n);
+                    size_t m = 0;
+                    while (m < rem && n < max_write) {
+                        auto s = std::min(rem - m, buf_size);
+                        v.emplace_back(iovec{ buf.get_write(), s});
+                        m += s;
+                    }
+                    auto s = co_await f.dma_write(max_size - rem, std::move(v), service::get_local_commitlog_priority());
+                    rem -= s;
+                }
+
+                co_await f.close();
```

Comment:
Avi mentioned that we need to fsync the file here, to sync file metadata once.

Reply:
close does sync + flush, i.e. fdatasync. I would assume this is enough, no?

Reply:
Do you mean seastar close doe it or syscall close does it? to my knowledge neither

Reply:
Oops, my intellisense tricked me. Looked at wrong close. Fixed.